File: orchestrator/src/linkup.ts
import { LinkupClient } from "linkup-sdk";
import { config } from "./config.js";

const client = new LinkupClient({ apiKey: config.linkupApiKey });

export type Depth = "fast" | "standard" | "deep";

export interface NormalizedResult {
  name: string;
  url: string;
  content: string;
}

// Raw search → normalized {name,url,content}[]. Handles both `{results:[...]}`
// and bare-array response shapes across SDK versions.
export async function linkupSearchResults(
  query: string,
  depth: Depth = "standard",
  opts: { maxResults?: number; fromDate?: Date; includeDomains?: string[] } = {},
): Promise<{ results: NormalizedResult[]; raw: unknown }> {
  const raw: any = await client.search({
    query,
    depth,
    outputType: "searchResults",
    ...(opts.maxResults ? { maxResults: opts.maxResults } : {}),
    ...(opts.fromDate ? { fromDate: opts.fromDate } : {}),
    ...(opts.includeDomains ? { includeDomains: opts.includeDomains } : {}),
  } as any);

  const list: any[] = Array.isArray(raw) ? raw : raw?.results ?? [];
  const results: NormalizedResult[] = list
    .filter((r) => r && (r.url || r.name))
    .map((r) => ({
      name: String(r.name ?? r.title ?? ""),
      url: String(r.url ?? ""),
      content: String(r.content ?? r.snippet ?? ""),
    }));
  return { results, raw };
}

// Structured extraction against a JSON schema (for SEO/pricing structured pulls).
export async function linkupStructured<T = unknown>(
  query: string,
  schema: object,
  depth: Depth = "deep",
): Promise<{ data: T; raw: unknown }> {
  const raw: any = await client.search({
    query,
    depth,
    outputType: "structured",
    structuredOutputSchema: schema,
  } as any);
  return { data: raw as T, raw };
}

File: agent/stalker/tools/linkup.py
"""Linkup live-search wrapper (the search power-up doing real work).
Normalizes response shapes across SDK versions to {name,url,content}."""
from __future__ import annotations

from typing import Any
from linkup import LinkupClient

from ..config import settings

_client: LinkupClient | None = None


def _get() -> LinkupClient:
    global _client
    if _client is None:
        _client = LinkupClient(api_key=settings.linkup_api_key)
    return _client


def search_results(
    query: str,
    depth: str = "standard",
    max_results: int = 8,
    from_date: str | None = None,
) -> tuple[list[dict[str, str]], Any]:
    """Run a Linkup searchResults query. Returns (normalized_results, raw)."""
    kwargs: dict[str, Any] = {
        "query": query,
        "depth": depth,
        "output_type": "searchResults",
    }
    raw = _get().search(**kwargs)

    # raw may be an object with `.results`, a dict, or a list.
    items = getattr(raw, "results", None)
    if items is None and isinstance(raw, dict):
        items = raw.get("results", [])
    if items is None and isinstance(raw, list):
        items = raw
    items = items or []

    out: list[dict[str, str]] = []
    for r in items:
        name = _attr(r, "name") or _attr(r, "title") or ""
        url = _attr(r, "url") or ""
        content = _attr(r, "content") or _attr(r, "snippet") or ""
        if url or name:
            out.append({"name": str(name), "url": str(url), "content": str(content)[:1200]})
    return out[:max_results], _to_jsonable(raw)


def _attr(obj: Any, key: str) -> Any:
    if isinstance(obj, dict):
        return obj.get(key)
    return getattr(obj, key, None)


def _to_jsonable(raw: Any) -> Any:
    """Best-effort convert the SDK response to a JSON-serializable structure."""
    if isinstance(raw, (dict, list, str, int, float, bool)) or raw is None:
        return raw
    for meth in ("model_dump", "dict"):
        fn = getattr(raw, meth, None)
        if callable(fn):
            try:
                return fn()
            except Exception:
                pass
    items = getattr(raw, "results", None)
    if items is not None:
        return {"results": [
            {"name": _attr(r, "name"), "url": _attr(r, "url"), "content": _attr(r, "content")}
            for r in items
        ]}
    return {"repr": str(raw)[:5000]}
